src/client/init/root.tsx

```diff
@@ -19,34 +19,53 @@ function useHydrated() {
 
 export const defineClientConfig = (config: RdtClientConfig) => config
 
-export const withDevTools = (Component: any, config?: RemixDevToolsProps) => () => {
-	hydrationDetector()
-	const hydrated = useHydrated()
-	if (!hydrated) return <Component />
-
-	return (
-		<>
-			<Component />
-			{createPortal(<RemixDevTools {...config} />, document.body)}
-		</>
-	)
-}
-
-/**
- *
- * @description Injects the dev tools into the Vite App, ONLY meant to be used by the package plugin, do not use this yourself!
- */
-export const withViteDevTools = (Component: any, config?: RemixDevToolsProps) => () => {
-	hydrationDetector()
-	function AppWithDevTools(props: any) {
+export const withDevTools =
+	(Component: any, config?: RemixDevToolsProps & { panelMode?: "auto" | "embedded" }) => () => {
+		hydrationDetector()
 		const hydrated = useHydrated()
+		const isEmbedded = useIsEmbeddedRDT()
 		if (!hydrated) return <Component />
+
+		if ((config?.panelMode ?? "auto") === "embedded") {
+			return (
+				<RDTEmbeddedContextProvider>
+					<Component />
+				</RDTEmbeddedContextProvider>
+			)
+		}
+
 		return (
 			<>
-				<Component {...props} />
-				{createPortal(<RemixDevTools {...config} />, document.body)}
+				<Component />
+				{!isEmbedded && createPortal(<RemixDevTools {...config} />, document.body)}
 			</>
 		)
 	}
-	return AppWithDevTools
-}
+
+/**
+ *
+ * @description Injects the dev tools into the Vite App, ONLY meant to be used by the package plugin, do not use this yourself!
+ */
+export const withViteDevTools =
+	(Component: any, config?: RemixDevToolsProps & { panelMode?: "auto" | "embedded" }) => () => {
+		hydrationDetector()
+		function AppWithDevTools(props: any) {
+			const hydrated = useHydrated()
+			if (!hydrated) return <Component />
+			if ((config?.panelMode ?? "auto") === "embedded") {
+				return (
+					<RDTEmbeddedContextProvider>
+						<Component />
+					</RDTEmbeddedContextProvider>
+				)
+			}
+
+			return (
+				<>
+					<Component {...props} />
+					{createPortal(<RemixDevTools {...config} />, document.body)}
+				</>
+			)
+		}
```

Comment:
I've been thinking, `RDTContextProvider` wraps both the EmbeddedDevTools and the regular ones, why wouldn't you just pop this panelMode option into that and here just not use the createPortal? instead of having 2 contexts?

Reply:
Then you in the vite plugin set it to embedded, that hides this createPortal, and wherever you use  <EmbeddedDevTools> you just pass in a prop that the panel is embedded (inside the EmbeddedDevTools to the context) and you get this behavior with less hassle and there is 1 context to read from

Reply:
Primarily because you really want React Context to be granular. One thing I was going to suggest to you was to separate the `dispatch` context out from the `state` context.

Having them both in the same context is inefficient, because every `useContext()` call subscribes the component to rerenders whenever the context value changes. That means that components that only need the `dispatch` will have to re-render whenever the `state` changes, even if they never read any `state`. By breaking those up into separate contexts, you get granular control of what causes a re-render. And since `dispatch` is stable, components that only need to dispatch won't re-render themselves after the state changes (often times in response to those components themselves invoking `dispatch`!).

With that in mind, having a separate context for "embedded" mode makes sense, since the value of this is essentially static, a component that needs to render differently based on embedded-mode or not won't need to re-render whenever the state managed by `RDTContextProvider` changes.

Reply:
I see, makes sense, okay